### PR TITLE
add host functions for cluster data (flash sale support)

### PR DESCRIPTION
**Pull request by localvar**
add host functions for accessing shared cluster data.
add cookbook for flash sale.
other enhancement.

Too many files are modified, propose to review commit by commit.

### CODE REVIEW

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 188 Deploy Test Success

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 188 Deploy Test Success

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 188 Deploy Test Success

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 188 Deploy Test Success

File: pkg/cluster/op.go [hunk 1]
```diff
@@ -173,3 +174,12 @@ func (c *cluster) GetRawPrefix(prefix string) (map[string]*mvccpb.KeyValue, erro
 
 	return kvs, nil
 }
+
+func (c *cluster) STM(apply func(concurrency.STM) error) error {
```
[the remark this answers is not in the document]
Comment (localvar) [reply pkg/cluster/op.go:178]: There are host functions like `hostClusterAddInterger`, which does a cluster-level atomic operation to increase/decrease an integer, and can be used to create a cluster-level counter, that's why I add the `STM` function.

File: doc/cookbook/flash_sale.md [hunk 2]
```diff
@@ -1,3 +1,484 @@
...
+		- [1.3 Verify What We Have Done](#13-verify-what-we-have-done)
+	- [2. Block All Requests Before Flash Sale Start](#2-block-all-requests-before-flash-sale-start)
+	- [3. Randomly Block Requests](#3-randomly-block-requests)
+	- [4. Lucky Once, Lucky Always](#4-lucky-once-lucky-always)
+	- [5. Limit The Number Of Permitted Users](#5-limit-the-number-of-permitted-users)
+	- [6. Reuse The Program](#6-reuse-the-program)
+		- [6.1 Parameters](#61-parameters)
+		- [6.2 Manage Shared Data](#62-manage-shared-data)
+
+A flash sale is a discount or promotion offered by an ecommerce store for a short period of time. The quantity is limited, which often means the discounts are higher or more significant than run-of-the-mill promotions.
```
[the remark this answers is not in the document]
Comment (localvar) [reply doc/cookbook/flash_sale.md]: updated

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 188 Deploy Test Success

File: pkg/api/wasm.go [hunk 3]
```diff
@@ -21,30 +21,163 @@ package api
...
 	"time"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/megaease/easegress/pkg/filter/wasmhost"
+	"go.etcd.io/etcd/client/v3/concurrency"
+	"gopkg.in/yaml.v2"
 )
 
-func (s *Server) postWasmCodeUpdateEvent() string {
+func (s *Server) isFilterExist(pipeline, filter, kind string) bool {
```
[the remark this answers is not in the document]
Comment (localvar) [reply pkg/api/wasm.go:34]: Yes, I agree with you that we should add something like `WASMController` in the future.

File: pkg/filter/wasmhost/vm.go [hunk 4]
```diff
@@ -133,14 +134,15 @@ func newWasmVM(engine *wasmtime.Engine, module *wasmtime.Module, params []string
...
 	chVM   chan *WasmVM
 	engine *wasmtime.Engine
 	module *wasmtime.Module
 	params []string
 }
 
-// NewWasmVMPool creates a wasm VM pool with 'size' VMs which execute 'code'
-func NewWasmVMPool(size int32, code []byte, params map[string]string) (*WasmVMPool, error) {
+// NewWasmVMPool creates a wasm VM pool according the spec of 'host' which execute 'code'
+func NewWasmVMPool(host *WasmHost, code []byte) (*WasmVMPool, error) {
```
[the remark this answers is not in the document]
Comment (localvar) [reply pkg/filter/wasmhost/vm.go:145]: Ok, we can do this later.

File: pkg/cluster/op.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (benja-wu) [reply pkg/cluster/op.go:178]: * I recommend using `Lock` for solving cluster-level atomic operations. But currently, the `lock` in the `storage` interface is whole Etcd level, so it's cool to introduce `STM` into Eg for this purpose. 
* But why not use at least the `Read commited` level's transaction in your implememtation?

[the remark this answers is not in the document]
Comment (localvar) [reply pkg/cluster/op.go:178]: I didn't notice we can set the isolation level :)

But after checking the ETCD source code, the default isolation level is `SerializableSnapshot`, which is correct, and `ReadCommited` does not fit our requirement.

[the remark this answers is not in the document]
Comment (benja-wu) [reply pkg/cluster/op.go:178]: Oh, I checked the `etcd/client/v3/concurrency` package, and noticed the `NewSTMReadCommitted` and ` NewSTMRepeatable` function family. So I thought the `NewSTM` is setting to `Read commited` level transaction. So I propose to use at least `NewSTMReadCommitted` instead. But I find out they are deprecated. 

Yes, currently `NewSTM` is actually `Serializable`'s transaction, which is far more strong than `Read committed`. So I am totally cool here.

**Review by benja-wu — APPROVED** [review]
LGTM

**Review by xxx7xxxx — APPROVED** [review]

**Pull request merged**